RFC-0021-Distributed-Pipeline-Parallel-Technical.md

```diff
@@ -0,0 +1,620 @@
+  x = layer2(x) # Assigned to processor 2
+  x = layer3(x) # Assigned to processor 3
+  x = layer4(x) # Assigned to processor 4
+  x = layer5(x) # Assigned to processor 5
+  return x
+```
+
+One way to implement this is to convert the code into a form that is programmatically manipulable (such as a `torch.fx` Graph), partition that IR such that the code for each stage resides in a separate program, and distribute that program to each of the processors. A runtime on each of those processors would then handle loading data (from a DataLoader or from a previous stage), running the partitioned code, calculating the loss (on the last stage), running backpropagation, and applying gradient updates. This is the [approach](https://github.com/microsoft/DeepSpeed/blob/af443f63f483f6ea6769b78b4b0f2407023e9aed/deepspeed/runtime/pipe/engine.py#L46) DeepSpeed takes in pipeline parallel execution (for example, [see](https://www.deepspeed.ai/tutorials/pipeline/) how you must pass your DataLoader and use an opaque `engine.train_batch` API in DeepSpeed, relinquishing control to the opaque runtime). However, this runtime essentially replaces the PyTorch training loop, and converting your training process to this scheme may be a burden for the end user (see Approach 4 for a comprehensive treatment of the design considerations for this approach).
+
+@zdevito proposed transparently [splitting a model into stages](https://colab.research.google.com/drive/1lGg2NqlvDwVmvBqejzni2yTmYE9rxfdr?usp=sharing) by employing a type of predicated execution[3]  For each stage, the Python code of the whole model is run, however operations outside of the ones relevant to each specific pipeline stage are “no-op”ed.
+
+![Predicated execution of a model in pipeline parallelism](https://i.imgur.com/opv5Wic.png)
+
+One thing not represented in this diagram is a 3rd dimension of time, i.e. there will be a “fill” stage where `forward` micro-batches fill in from left-to-right, and a drain stage where `backward` micro-batches drain out of the pipeline from right-to-left. The diagram above represents the pipeline in a “steady-state” condition.
+
+
+### Expanding Predication to Include the Training Loop
+
+The next logical step is to expand this proposal to encompass the training loop, allowing the user to continue to write their arbitrary Python training loop, but overlaying pipeline parallelism semantics on this using predicated execution.
+
+We can pull out the parts of the canonical training loop from the beginning and investigate how they should be executed under a predicated, pipeline parallel training loop:
+
+**Data Loader**
+
+The data loader should only load input data on rank 0. We can view this as predicating `true` on rank 0 and false on all other ranks. We can also commingle this with `recv` for stages != 0. i.e. under the hood the dataloader object will return an input micro-batch on rank 0, but will return the intermediate value received from `rank - 1` on all ranks != 0.
```

Comment:
@thomasw21 if this is something that is an expected use case, we should discuss this further as it has broader implications on the rest of the RFC and we should likely re-do all of the designs.

To help me understand: what would the execution pattern look like when pipelining a non-sequential (arbitrary DAG) model? In particular, can there exist disconnected nodes  which must be scheduled somehow? For example:

```
def forward(self, x):
  y = torch.zeros(3, 4) # On forward, must be scheduled for execution independently of the function inputs
  z = self.layer_1(x) # rank 0
  a = self.layer_2(z + y) # rank 1
  return a
```

(Hypothetically we could also have some sort of sink node in the forward that results in a disconnected edge in the autograd graph, but I can't think of a concrete scenario where that would happen)

Reply:
I would expect that disconnected graphs should work. Typically when using absolute positional embeddings, we obtain the position embeddings like the following:

```
def forward(self, x):
   pos_emb = self.get_position_embeddings() # returns nn.Parameter
   token_emb = self.compute_token_embeddings(x)
   return pos_emb + token_emb
```

However now that I think about it, it's highly unlikely that there exist a rank that holds a parentless node (otherwise you would just fuse it with one of its children, the first one in the rank ordering, and use that as a proxy parent):


```
pos_emb -------> output
             /
token_emb --
```

to

```
token_emb ----> pos_emb + token_emb ----> output
```

The reason I think of DAG, is because let's say you have ensembling techniques, where you have 5 experts, which are then aggregated. I'd expect the 5 experts to be able to run in parallel. Now imagine each expert is held inside one rank each. You'd have something like 6 ranks (5 + 1 to run experts and then aggregate).

Current framework would work like this in my understanding:
```
expert 0 ----> expert 1 + expert 0 ----> expert 2 + expert 1 + expert 0 ---> etc 
```

Each rank would actually hold its own expert, run that expert, but actually keep forwarding previous expert results to next rank no? This means that fundamentally we should only be able to run that in a sequential fashion? I would have expected all experts to be able to run in parallel. The agregator can still require to receive values from its experts in an arbitrary order (but that means we're bottlenecked by the slowest expert, instead of the sum all experts).

Does that make sense?

Reply:
@thomasw21 I think modeling arbitrary graph programs in a pipeline parallel API is certainly possible (e.g. it was done in the `fairscale` experimental distributed pipeline prototype https://github.com/facebookresearch/fairscale/blob/7c2c3e0046eb3a1b1c98fd24f61fb09f90f6817f/fairscale/experimental/nn/distributed_pipeline/graph.py#L56).

OTOH, I would ask whether we can decompose this problem into simpler solutions that can be composed with each other. For example, in the first example, could we group the `get_position_embeddings()` call and the `compute_token_embeddings` call into a single pipeline stage and use RPCs to run those in parallel on remote hosts. Similarly, for an MoE architecture, can we commingle each expert into a single pipeline stage, but within that pipeline stage use remote RPCs to run the experts on remote hosts in parallel?

Reply:
Won't that be incredible inefficient speed-wise comparatively to other stages with a lot more logic in them? Also wrt to memory usage - as you will want to partition each stage so that memory usage will be balanced.

Reply:
@stas00 could you clarify? Are you talking about the examples specifically or the general concept of calling RPCs within a pipeline stage?

In terms of latency, it should be equivalent to the critical path through the (2 or 5) parallel calls plus some RPC overhead. If these things are actually running on the same box, you could also just issue those two commands on different CUDA streams and get parallelism that way

Reply:
I'm thinking in the terms of pipeline stages as they are currently used by Megatron and Deepspeed.

A typical pipeline for a transformer block is just one or more transformer blocks with an appendix of the embedding in the first pipeline stage, e.g.:

```
[emb|trans|trans] -> [trans|trans] -> [trans|trans|embed]
```
which ensures that they are mostly of the same size and do about the same amount of processing.

but if we now introduce a much finer granularity stages such as a stage that computes `get_position_embeddings` which is super quick and requires almost no memory I'm no longer sure what happens.

Unless your definition of a pipeline stage is something else.

Reply:
@stas00 agreed that splitting those embeddings functions off into stages is dubious, but I was going off of @thomasw21's original formulation, so I would extend that question to him.

Reply:
Yeah sorry, when mentioning embeddings I was trying to think of examples of parentless nodes. I think that it's highly unlikely that you would create a pipeline stage that doesn't require a previous stage to be ran before or depend on any inputs.

I think the part I'm perhaps a bit unfamiliar is the concept of having multiple hosts on a single pipeline stage. I'm mostly familiar with deepspeed's version. Stages are assigned a single device. We mainly get parallel execution thanks to microbatching/scheduling.

Taking the multi experts again. If I understand correctly there'd be two stages, one to run all the experts, and another one to run the aggregator? If so that makes sense to me. The rules should become like:
 - Within a single pipeline stage, multiple hosts run in parallel.
 - Ranks feed their results sequentially to next ranks.
 
But then if you manually have to setup RPCs, aren't you manually doing Pipeline parallelism? ie the whole model is a single stage and use RPCs to schedule calls on different devices?

Reply:
@thomasw21 here's what i'm describing, maybe you have something else in mind:

Given the definitions of pipelining used in the literature, pipeline parallelism looks like turning the following:

```
def my_model(x):
  x = op1(x)
  x = op2(x)
  x = op3(x)
  return x
```

Into something like (suppose we cut between op2 and op3 into 2 stages):

```
def my_model0(x):
  x = op1(x)
  x = op2(x)
  return x

def my_model1(x):
  x = op3(x)
  return x
```

And then the composition `my_model1(my_model0(x))` would run in pipelined execution. I think in the MoE example, the original source would look something like:

```
def my_model(x):
  x = op1(x)
  x = op2(x)
  x = op3(x)

  gates = gate_fn(x)
  moe_results = []
  for i, expert in enumerate(experts):
    moe_results.append(expert(x[i]))
  x = combine(moe_experts)
  return x
```

Suppose we apply the asme pipelining policy, i.e.:

```
def my_model0(x):
  x = op1(x)
  x = op2(x)
  return x

def my_model1(x):
  x = op3(x)

  gates = gate_fn(x)
  moe_results = []
  for i, expert in enumerate(experts):
    moe_results.append(expert(x[i]))
  x = combine(moe_results)
  return x
```

As you point out, there is inherent parallelism in the parallel application of the experts, i.e. in the `enumerate(experts)` loop. However, with respect to how the code appears on the screen, this is "horizontal" parallelism vs. the "vertical" parallelism of pipelining. So my proposal is to basically implement horizontal parallelism in a way that is orthogonal to pipelining, i.e.:

```
def my_model0(x):
  x = op1(x)
  x = op2(x)
  return x

def my_model1(x):
  x = op3(x)

  gates = gate_fn(x)
  futures = []
  for i, expert in enumerate(experts):
    futures.append(rpc(expert(x[i])))

  moe_results = []
  for fut in futures:
    moe_results = fut.wait()

  x = combine(moe_results)
  return x
```

In this way, horizontal parallelism is achieved while keeping the pipeline parallel system simple and focused on vertical parallelism. The same argument can be made for model parallelism, i.e. pipeline parallelism can operate on ShardedTensor instances that dispatch to collectives, but pipeline parallelism itself does not need to be aware of this scheme

Reply:
@thomasw21 Put another way, pipeline parallelism is about parallelizing code _across micro-batches_, whereas the MoE parallelism is _within a micro-batch_, thus I think these should be orthogonal to each other. Does that make sense?

Reply:
I see, this makes sense. Would the partitioning algorithm be able to work orthogonally with having these kind of systems but in place?

Reply:
To support the discussion of using MoE in the pipeline

Microsoft and Google have just published architectures that use MoE:

https://ai.googleblog.com/2021/12/more-efficient-in-context-learning-with.html
https://www.deepspeed.ai/news/2021/12/09/deepspeed-moe-nlg.html

Deepspeed's version uses Megatron-Deepspeed so it's already based on Deepspeed PP implementation - so you could probably check how they dealt with MoE - the Moe Branch is here: https://github.com/microsoft/Megatron-DeepSpeed/tree/moe-training

But reading closer Deepspeed's version uses only top-1 expert, so it's then irrelevant to this discussion, as it runs only 1 MoE.

Google's arch uses 2 MoEs and thus is relevant as it needs to parallelize the processing of these 2.